New python Solution USACO January 2019 Bronze P2 (#1400)
cpinitiative/usaco-guide: solutions/USACO feb21-year-cow.mdx

---
id: usaco-1107
source: USACO Bronze 2021 February
title: Year of the Cow
author: Ananth Kashyap
---

[Official Analysis](http://www.usaco.org/current/data/sol_prob1_bronze_feb21.html)

> In this problem, we want to find the number of years separating Elsie and Bessie. For every line of the input, we can calculate the number of years separating the two cows. Then, we can store a cumulative value that holds the number of years separating Bessie and the second cow in the input line, as the first cow in the input line is guaranteed to appear in a previous line. When, Elsie appears (as the second cow) in the input line, we can readjust our cumulative value, which will be our answer.

Reviewer: ```suggestion
In this problem, we want to find the number of years separating Elsie and Bessie. For every line of the input, we can calculate the number of years separating the two cows. Then, we can store a cumulative value that holds the number of years separating Bessie and the second cow in the input line, as the first cow in the input line is guaranteed to appear in a previous line. When Elsie appears (as the second cow) in the input line, we can readjust our cumulative value, which will be our answer.
```

## Implementation

**Time Complexity:** $\mathcal{O}(N^2)$

<LanguageSection>

<PySection>

> ```python

Reviewer: use tabs instead of spaces to indent

> n = int(input(''))

Reviewer: ```suggestion
n = int(input())
```

arr = []
for i in range(n):
    # Keep track of only keywords from each line of the input

>     tarr = input('').split(' ')

Reviewer: ```suggestion
    tarr = input().split(' ')
```

    arr.append([tarr[0], tarr[3], tarr[4], tarr[7]])

> # Store the order of the Chinese New Year names in a new variable

Reviewer: ```suggestion
# Store the order of the Chinese year names
```

years = ['Ox', 'Tiger', 'Rabbit', 'Dragon', 'Snake', 'Horse',
         'Goat', 'Monkey', 'Rooster', 'Dog', 'Pig', 'Rat']

for i in range(n):

>     # Find out the number of years seperating the cows in each line of the input

Reviewer: ```suggestion
    # Find the number of years separating the cows in each line of the input
```

    if arr[i][3] == 'Bessie':
        arr[i].append(years.index(arr[i][2]))
        arr[i].append(0)
    else:
        for j in range(i - 1, -1, -1):
            if arr[j][0] == arr[i][3]:
                value = j
        arr[i].append(years.index(arr[i][2]))
        arr[i].append(years.index(arr[value][2]))
    if arr[i][1] == 'previous':
        if arr[i][4] >= arr[i][5]:
            arr[i].append(-12 + arr[i][4] - arr[i][5])
        else:
            arr[i].append(arr[i][4] - arr[i][5])
    else:
        if arr[i][4] <= arr[i][5]:
            arr[i].append(12 - (arr[i][5] - arr[i][4]))
        else:
            arr[i].append(arr[i][4] - arr[i][5])

>     # Store the number of years seperating Bessie and the second cow in the line

Reviewer: ```suggestion
    # Store the number of years separating Bessie and the second cow in the line
```
